crates/router/src/connector/square.rs

```diff
@@ -809,24 +811,93 @@ impl ConnectorIntegration<api::RSync, types::RefundsData, types::RefundsResponse
 
 #[async_trait::async_trait]
 impl api::IncomingWebhook for Square {
-    fn get_webhook_object_reference_id(
+    fn get_webhook_source_verification_algorithm(
         &self,
         _request: &api::IncomingWebhookRequestDetails<'_>,
+    ) -> CustomResult<Box<dyn common_utils::crypto::VerifySignature + Send>, errors::ConnectorError>
+    {
+        Ok(Box::new(common_utils::crypto::HmacSha256))
+    }
+
+    fn get_webhook_source_verification_signature(
+        &self,
+        request: &api::IncomingWebhookRequestDetails<'_>,
+    ) -> CustomResult<Vec<u8>, errors::ConnectorError> {
+        let encoded_signature =
+            super_utils::get_header_key_value("x-square-hmacsha256-signature", request.headers)?;
+        let signature = consts::BASE64_ENGINE
+            .decode(encoded_signature)
+            .into_report()
+            .change_context(errors::ConnectorError::WebhookSignatureNotFound)?;
+        Ok(signature)
+    }
+
+    fn get_webhook_source_verification_message(
+        &self,
+        request: &api::IncomingWebhookRequestDetails<'_>,
+        _merchant_id: &str,
+        _secret: &[u8],
+    ) -> CustomResult<Vec<u8>, errors::ConnectorError> {
+        Ok(format!("{}{}", request.uri, String::from_utf8_lossy(request.body)).into_bytes())
+    }
+
+    fn get_webhook_object_reference_id(
+        &self,
+        request: &api::IncomingWebhookRequestDetails<'_>,
     ) -> CustomResult<api::webhooks::ObjectReferenceId, errors::ConnectorError> {
-        Err(errors::ConnectorError::WebhooksNotImplemented).into_report()
+        let webhook_body: square::SquareWebhookBody = request
+            .body
+            .parse_struct("SquareWebhookBody")
+            .change_context(errors::ConnectorError::WebhookReferenceIdNotFound)?;
+
+        match webhook_body.data.object {
+            square::SquareWebhookObject::Payment(_) => {
+                Ok(api_models::webhooks::ObjectReferenceId::PaymentId(
+                    api_models::payments::PaymentIdType::ConnectorTransactionId(
+                        webhook_body.data.id,
+                    ),
+                ))
+            }
+            square::SquareWebhookObject::Refund(_) => {
+                Ok(api_models::webhooks::ObjectReferenceId::RefundId(
+                    api_models::webhooks::RefundIdType::ConnectorRefundId(webhook_body.data.id),
```

Comment:
do we get back our refund_id in web hooks? if so let's use that

Reply:
No. We get the refund id assigned by connector square as part of Webhooks. Our ID is just used as the Idempotency key.